File: reader/DNAVisGroup.cpp
#include "DNAVisGroup.h"

DNAVisGroup::DNAVisGroup(string name) : DNAGroup(name) {}
DNAVisGroup::~DNAVisGroup(void) {}

void DNAVisGroup::make_from_dgi(DatagramIterator& dgi, DNAStorage* store)
{
	DNAGroup::make_from_dgi(dgi, store);

	unsigned short num_edges = dgi.get_uint16();
	for (unsigned short i = 0; i < num_edges; i++)
	{
		unsigned short start_point = dgi.get_uint16();
		unsigned short end_point = dgi.get_uint16();
		add_suit_edge(store->get_suit_edge(start_point, end_point));
	};

	unsigned short num_visibles = dgi.get_uint16();
	for (unsigned short i = 0; i < num_visibles; i++)
	{
		add_visible(dgi_extract_string8(dgi));
	};

	unsigned short num_cells = dgi.get_uint16();
	for (unsigned short i = 0; i < num_cells; i++)
	{
		unsigned char w = dgi.get_uint8();
		unsigned char h = dgi.get_uint8();

		int x = dgi.get_int32();
		int y = dgi.get_int32();
		int z = dgi.get_int32();

		DNABattleCell c = DNABattleCell();
		c.set_width_height(w, h);
		c.set_pos(new LVector3f(x / 100.0, y / 100.0, z / 100.0));
		add_battle_cell(c);
	};
};

void DNAVisGroup::add_battle_cell(DNABattleCell cell)
{
	m_cells.push_back(cell);
};

void DNAVisGroup::add_suit_edge(DNASuitEdge edge)
{
	m_edges.push_back(edge);
};

void DNAVisGroup::add_visible(string visible)
{
	m_visibles.push_back(visible);
};

size_t DNAVisGroup::get_num_battle_cells()
{
	return m_cells.size();
};

size_t DNAVisGroup::get_num_suit_edges()
{
	return m_edges.size();
};

size_t DNAVisGroup::get_num_visibles()
{
	return m_visibles.size();
};

DNABattleCell DNAVisGroup::get_battle_cell(size_t index)
{
	return m_cells.at(index);
};

DNASuitEdge DNAVisGroup::get_suit_edge(size_t index)
{
	return m_edges.at(index);
};

string DNAVisGroup::get_visible_name(size_t index)
{
	return m_visibles.at(index);
};

void DNAVisGroup::traverse(NodePath& np, DNAStorage* store)
{
	DNAGroup::traverse(np, store);
};

DNAVisGroup* DNAVisGroup::get_vis_group()
{
	return this;
};

File: libpandadna/DNAVisGroup.py
import DNAGroup, DNASuitEdge, DNABattleCell

from common import *

class DNAVisGroup(DNAGroup.DNAGroup):
    COMPONENT_CODE = 2

    def __init__(self, name):
        DNAGroup.DNAGroup.__init__(self, name)

        self.visibles = []
        self.suitEdges = []
        self.battleCells = []

    def getVisGroup(self):
        return self

    def addVisible(self, visible):
        self.visibles.append(visible)

    def addSuitEdge(self, suitEdge):
        self.suitEdges.append(suitEdge)

    def addBattleCell(self, battleCell):
        self.battleCells.append(battleCell)
    
    def getBattleCell(self, i):
        return self.battleCells[i]

    def getNumBattleCells(self):
        return len(self.battleCells)

    def getNumSuitEdges(self):
        return len(self.suitEdges)

    def getNumVisibles(self):
        return len(self.visibles)

    def getSuitEdge(self, i):
        return self.suitEdges[i]

    def getVisibleName(self, i):
        return self.visibles[i]

    def removeBattleCell(self, cell):
        self.battleCells.remove(cell)

    def removeSuitEdge(self, edge):
        self.suitEdges.remove(edge)

    def removeVisible(self, visible):
        self.visibles.remove(visible)

    def makeFromDGI(self, dgi, store):
        DNAGroup.DNAGroup.makeFromDGI(self, dgi)
        
        numEdges = dgi.getUint16()
        for i in xrange(numEdges):
            index = dgi.getUint16()
            endPoint = dgi.getUint16()
            self.addSuitEdge(store.getSuitEdge(index, endPoint))
                
        numVisibles = dgi.getUint16()
        for i in xrange(numVisibles):
            self.addVisible(dgi_extract_string8(dgi))
                
        numCells = dgi.getUint16()
        for i in xrange(numCells):
            w = dgi.getUint8()
            h = dgi.getUint8()
            
            x, y, z = [dgi.getInt32() / 100.0 for _ in xrange(3)]
            
            self.addBattleCell(DNABattleCell.DNABattleCell(w, h, (x, y, z)))
            